File: api-python/services/search_engine.py
import asyncio
import uuid
from datetime import datetime, timezone

from database import create_search, create_result, get_search, update_search, update_result
from services.places import search_text
from services.custom_search import verify_website


async def run_search(query: str, location: str, radius: int = 5000, type: str | None = None,
                     search_id: str | None = None) -> dict:
    if search_id is None:
        search_id = str(uuid.uuid4())
        create_search(search_id, query, location, radius, type)
    update_search(search_id, status='processing')

    try:
        places = await search_text(query, location, radius)

        with_website = 0
        without_website = 0
        pending_verification = 0
        result_ids: dict[str, str] = {}

        for place in places:
            if place.get('website'):
                with_website += 1
                continue

            result_id = str(uuid.uuid4())
            create_result(
                id=result_id, search_id=search_id,
                name=place['name'], address=place.get('address'),
                phone=place.get('phone'), place_id=place.get('placeId'),
                website=None,
                status='pending_verification', verification_method='google_maps',
                rating=place.get('rating'), total_ratings=place.get('totalRatings'),
                categories=place.get('categories'), lat=place.get('lat'), lng=place.get('lng'),
            )
            result_ids[place['placeId']] = result_id
            pending_verification += 1

        verifiable = [p for p in places if not p.get('website')]

        for place in verifiable:
            try:
                verification = await asyncio.to_thread(verify_website, place['name'], location)
                rid = result_ids.get(place['placeId'])
                if not rid:
                    continue

                if verification['hasWebsite']:
                    site_url = verification['matches'][0]['link'] if verification['matches'] else None
                    update_result(
                        rid,
                        website=site_url,
                        website_verified=1,
                        website_confirmed=0,
                        status='has_website',
                        verification_method='google_search',
                    )
                    pending_verification -= 1
                    with_website += 1
                elif verification.get('networkSite'):
                    update_result(
                        rid,
                        network_site=verification['networkSite'],
                        website_verified=1,
                        website_confirmed=0,
                        status='no_website',
                        verification_method='google_search',
                    )
                    pending_verification -= 1
                    without_website += 1
                else:
                    update_result(
                        rid,
                        website_verified=1,
                        website_confirmed=1,
                        status='no_website',
                        verification_method='google_search',
                    )
                    pending_verification -= 1
                    without_website += 1
            except Exception:
                rid = result_ids.get(place['placeId'])
                if rid:
                    update_result(
                        rid,
                        status='no_website',
                        verification_method='google_maps',
                        notes='Google Search verification failed, defaulting to no_website',
                    )
                pending_verification -= 1
                without_website += 1

            update_search(search_id,
                          with_website=with_website,
                          without_website=without_website,
                          pending_verification=pending_verification)

        now = datetime.now(timezone.utc).isoformat()
        update_search(search_id,
                      status='completed',
                      total_found=len(places),
                      with_website=with_website,
                      without_website=without_website,
                      pending_verification=pending_verification,
                      completed_at=now)

        return get_search(search_id)

    except Exception as err:
        update_search(search_id, status='failed', error=str(err))
        raise

File: api/src/services/searchEngine.js
import { v4 as uuid } from 'uuid';
import { createSearch, updateSearch, getSearch } from '../db/searches.js';
import { createResult, updateResult, getResult } from '../db/results.js';
import { searchText } from './places.js';
import { verifyWebsite } from './customSearch.js';

export async function runSearch({ query, location, radius = 5000, type }) {
  const id = uuid();
  createSearch({ id, query, location, radius, type });
  updateSearch(id, { status: 'processing' });

  try {
    const places = await searchText(query, location, radius);

    let withWebsite = 0;
    let withoutWebsite = 0;
    let pendingVerification = 0;

    const resultIds = {};

    for (const place of places) {
      if (place.website) {
        withWebsite++;
        continue;
      }

      const resultId = uuid();
      createResult({
        id: resultId,
        searchId: id,
        ...place,
        status: 'pending_verification',
        verificationMethod: 'google_maps',
      });
      resultIds[place.placeId] = resultId;
      pendingVerification++;
    }

    const verifiable = places.filter(p => !p.website);
    for (const place of verifiable) {
      try {
        const verification = await verifyWebsite(place.name, location);
        const resultId = resultIds[place.placeId];
        if (!resultId) continue;

        if (verification.hasWebsite) {
          const siteUrl = verification.matches[0]?.link || null;
          updateResult(resultId, {
            website: siteUrl,
            website_verified: 1,
            website_confirmed: 0,
            status: 'has_website',
            verification_method: 'google_search',
          });
          pendingVerification--;
          withWebsite++;
        } else if (verification.networkSite) {
          updateResult(resultId, {
            network_site: verification.networkSite,
            website_verified: 1,
            website_confirmed: 0,
            status: 'no_website',
            verification_method: 'google_search',
          });
          pendingVerification--;
          withoutWebsite++;
        } else {
          updateResult(resultId, {
            website_verified: 1,
            website_confirmed: 1,
            status: 'no_website',
            verification_method: 'google_search',
          });
          pendingVerification--;
          withoutWebsite++;
        }
      } catch {
        const resultId = resultIds[place.placeId];
        if (resultId) {
          updateResult(resultId, {
            status: 'no_website',
            verification_method: 'google_maps',
            notes: 'Google Search verification failed, defaulting to no_website',
          });
        }
        pendingVerification--;
        withoutWebsite++;
      }

      updateSearch(id, {
        with_website: withWebsite,
        without_website: withoutWebsite,
        pending_verification: pendingVerification,
      });
    }

    updateSearch(id, {
      status: 'completed',
      total_found: places.length,
      with_website: withWebsite,
      without_website: withoutWebsite,
      pending_verification: pendingVerification,
      completed_at: new Date().toISOString(),
    });

    return getSearch(id);
  } catch (err) {
    updateSearch(id, { status: 'failed', error: err.message });
    throw err;
  }
}
